### ISSUE TITLE
[FP]: PC0030 - Skip SetLoadFields when record is set into another record.

### ISSUE THREAD

**Issue by guidorobben** [ALCops/Analyzers#260]
### Analyzer

PlatformCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

PC0030

### Description

Can we ignore PC0030 when we set the record into another record? In this case I'm not going to use SetLoadFields.

<img width="1639" height="460" alt="Image" src="https://github.com/user-attachments/assets/a0151a54-365b-435b-9aad-38ac4c5e5195" />

### AL code to reproduce

```al

```

**Comment by AnsBarMIT** [comment]
Just had a similar issue, i was thinking maybe an explicit `SetLoadFields()` without parameters (and potentially an explaining comment) to indicate loading all fields is wanted here could also be recognized by the rule?

**Comment by Arthurvdv** [comment]
EDIT: Replaced with updated information from #261

> We have some code that copies records from a local variable to a temporary table. According to the documentation, partial records should not be used here [(MS Learn)](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-partial-records):
> 
> > For performance reasons, it's not recommended to use partial records on a record that will do inserts, deletes, renames, field transfers, or copies to temporary records.

This is particularly helpful @AnsBarMIT!

I think I can detect if the variable is assigned to another (temporary) record and suppress the diagnostic. While the documentation states explicitly temporary records, I think it for the PC0030 diagnostic we should also suppress the diagnostic on non-temporary records.

**Comment by guidorobben** [comment]
the comment @AnsBarMIT  made is a good one I think.  Once the empty SetLoadFields is there, the rule should be suppressed. 

from https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/record/record-setloadfields-method

Remarks
Calling SetLoadFields on a record without passing any fields will reset the fields selected to load to the default, where all readable normal fields are selected for load.

**Comment by Arthurvdv** [comment]
I'm doubting on the _SetLoadFields on a record without passing any fields_ part here.

### Example A

In this example we raise the diagnostic on the second `MyTable.FindSet()` as it's now missing a SetLoadFields.

```AL
procedure MyProcedure()
var
    MyTable: Record MyTable;
begin
    MyTable.SetLoadFields(MyField);
    if MyTable.FindSet() then
        repeat
        // My Business Logic here...
        until MyTable.Next() = 0;
    Clear(MyTable);
    // My Business Logic here...

    if MyTable.FindSet() then // PC0030: Use SetLoadFields before 'MyTable.FindSet()' to improve performance by loading only the fields that are needed.
        repeat
        // My Business Logic here...
        until MyTable.Next() = 0;
end;
```

### Example B

I would like to be able to raise a diagnostic on the scenario below, where we would probably want a SetLoadFields on the second `MyTable.FindSet()` as it will now behave the same as a missing SetLoadFields from example A.

```AL
procedure MyProcedure()
var
    MyTable: Record MyTable;
begin
    MyTable.SetLoadFields(MyField);
    if MyTable.FindSet() then
        repeat
        // My Business Logic here...
        until MyTable.Next() = 0;
    MyTable.SetLoadFields(); // This line is ambiguous between 'remove the previous SetLoadFields' and 'I want to explicitly set the SetLoadFields to load all the fields' for the upcoming FindSet

    // My Business Logic here...

    if MyTable.FindSet() then // Do we expect an diagnoastic here?
        repeat
        // My Business Logic here...
        until MyTable.Next() = 0;
end;
```

The ambiguous of the _SetLoadFields on a record without passing any fields_ is a challenge I believe. I'm open for suggestions here 🙋

-----

Back on topic a bit, the _Skip SetLoadFields when record is set into another record_ is less complex to resolve I believe, let  me investigate that one.

**Comment by guidorobben** [comment]
I agree with example A. In that scenario I dont mind adding a empty SetLoadFields above the second findset

About B.

when you have a commit, you must add behind it for a comment or above it. We have the rule to put the setloadfields the line before the find. 

Cant we just expect the existence of the setloadfields line above the Find. Unless its not already set? 

```
procedure MyProcedure()
var
    MyTable: Record MyTable;
begin
    MyTable.SetLoadFields(MyField);
    if MyTable.FindSet() then
        repeat
        // My Business Logic here...
        until MyTable.Next() = 0;
    MyTable.SetLoadFields(); // Ignore this line!!!

    // My Business Logic here...

    MyTable.SetLoadFields(); // Expect a SetLoadFields Line here
    if MyTable.FindSet() then // Do we expect an diagnoastic here?
        repeat
        // My Business Logic here...
        until MyTable.Next() = 0;
end;
```

**Comment by TKapitan** [comment]
Yes, I see this problem across multiple projects when assigning the variable to a temporary record.

**Comment by Arthurvdv** [comment]
In the upcoming 0.8.1 release of ALCops, I’ve addressed the false positive related to _“when a record is assigned to another record”_.

For now, I’m leaving the explicit SetLoadFields() call without parameters out of scope, since handling that scenario adds a lot of complexity.

**Comment by Arthurvdv** [comment]
@guidorobben and @TKapitan, with the release of version `v0.8.1` of ALCops this issue should now be resolved I believe.

I’m going to close this issue now, but if you have any feedback or run into bugs/false positives, feel free to reopen it (or create a new issue).
